src/uct/ib/rdmacm/rdmacm_cm.c

```diff
@@ -65,52 +67,272 @@ ucs_status_t uct_rdmacm_cm_reject(struct rdma_cm_id *id)
                   ibv_get_device_guid(verbs->device), &ret);
     if (ret == -1) {
-        ucs_error("cm %p: cannot allocate hash entry for CQ", cm);
-        return UCS_ERR_NO_MEMORY;
+        ucs_error("cm %p: cannot allocate hash entry for qp context", cm);
+        status = UCS_ERR_NO_MEMORY;
+        goto out;
     }
 
     if (ret == 0) {
         /* already exists so use it */
-        cq = kh_value(&cm->cqs, iter);
+        ctx = kh_value(&cm->ctxs, iter);
     } else {
-        /* Create a dummy completion queue */
-        cq = ibv_create_cq(verbs, 1, NULL, NULL, 0);
-        if (cq == NULL) {
-            kh_del(uct_rdmacm_cm_cqs, &cm->cqs, iter);
-            ucs_error("ibv_create_cq() failed: %m");
-            return UCS_ERR_IO_ERROR;
+        /* Create a qp context */
+        ctx = ucs_malloc(sizeof(struct uct_rdmacm_cm_qp_context),
+                         "qp_context");
+        if (ctx == NULL) {
+            ucs_error("cm %p: malloc qp context failed: %m", cm);
+            status = UCS_ERR_NO_MEMORY;
+            goto err_kh_del;
         }
 
-        kh_value(&cm->cqs, iter) = cq;
+        uct_rdmacm_cm_query_reserved_qpn_cap(verbs, ctx);
+        if (ctx->reserved_qpn_support) {
+            uct_rdmacm_cm_query_reserved_qpn_cap_2(verbs, ctx);
+            pthread_mutex_init(&ctx->reserved_qpn_ctx.mutex, NULL);
+            ucs_list_head_init(&ctx->reserved_qpn_ctx.blk_list);
+        } else {
+            /* Create a dummy completion queue */
+            ctx->dummy_qp_ctx.cq = ibv_create_cq(verbs, 1, NULL, NULL, 0);
+            if (ctx->dummy_qp_ctx.cq == NULL) {
+                ucs_error("ibv_create_cq() failed: %m");
+                status = UCS_ERR_IO_ERROR;
+                goto err_free_ctx;
+            }
+        }
+
+        kh_value(&cm->ctxs, iter) = ctx;
     }
 
-    *cq_p = cq;
+    *ctx_p = ctx;
     return UCS_OK;
+
+err_free_ctx:
+    ucs_free(ctx);
+err_kh_del:
+    kh_del(uct_rdmacm_cm_qp_contexts, &cm->ctxs, iter);
+out:
+    return status;
 }
 
-void uct_rdmacm_cm_cqs_cleanup(uct_rdmacm_cm_t *cm)
+ucs_status_t
+uct_rdmacm_cm_reserved_qpn_blk_alloc(uct_rdmacm_cm_t *cm,
+                                     struct ibv_context *verbs,
+                                     struct uct_rdmacm_cm_qp_context *ctx,
+                                     struct uct_rdmacm_cm_reserved_qpn_blk **blk_p)
 {
-    struct ibv_cq *cq;
-    int ret;
+    ucs_status_t status = UCS_ERR_UNSUPPORTED;
+
+#if HAVE_DECL_MLX5DV_IS_SUPPORTED
+    char in[UCT_IB_MLX5DV_ST_SZ_BYTES(create_reserved_qpn_in)]   = {};
+    char out[UCT_IB_MLX5DV_ST_SZ_BYTES(general_obj_out_cmd_hdr)] = {};
+    struct uct_rdmacm_cm_reserved_qpn_blk *blk;
+    void *attr;
+    uint32_t qpns_per_obj;
+
+    qpns_per_obj = uct_rdmacm_cm_get_qpns_per_obj(ctx->reserved_qpn_ctx);
+    if (qpns_per_obj > UCT_RDMACM_RESERVED_QPN_BLK_BMP_BITS) {
+        ucs_error("reserved qpn bit map length is not enough");
+        return UCS_ERR_INVALID_PARAM;
+    };
```

Comment:
we could check this during device context initialization 
also, if qpns_per_obj is larger than bitmap size, we could ignore the extra objects instead of failing?

Reply:
@yosefe We can do it, but maybe we don't want.  The bitmap size is not enough is unexpected. It will happen when the FW implementation is changed. I think we prefer to notice it and change UCX code (make the bitmap size enough) according to the FW change. No purpose to not use full block objects.

Reply:
so it should be warning, not error, and use smaller size. so user application can continue working while we fix the problem.

Reply:
we can remove the bitmap (track by refcount) so this will no longer be relevant

Reply:
@yosefe Fixed.